### ISSUE TITLE
baseline for long-tail cifar10 is 77.47%

### ISSUE THREAD

**Issue by hbsz123** [richardaecn/class-balanced-loss#5]
![image](https://user-images.githubusercontent.com/18299981/62186739-d1a30200-b399-11e9-85b8-6d8c77ec9374.png)

Because my implementation for cb-focal in pytorch can only reach 77% accuracy for long-tail cifar10. I have ran your source code without any change. But the baseline is 77.5% (My baseline for pytorch in actually 75%.), which is almost 2.7% higher than the results reported in your paper. 
Does this mean cb-focal is only 2% higher than baseline?

**Comment by richardaecn** [comment]
Hi @hbsz123, from the screenshot, it looks like you are using softmax, not class-balanced focal loss. For the performance on CIFAR-10, please refer to Figure 8 of Appendix A in the paper (https://arxiv.org/abs/1901.05555).

**Comment by hbsz123** [comment]
> Hi @hbsz123, from the screenshot, it looks like you are using softmax, not class-balanced focal loss. For the performance on CIFAR-10, please refer to Figure 8 of Appendix A in the paper (https://arxiv.org/abs/1901.05555).

Thanks for your kind reply. 
I use softmax to train actually.I found that the baseline trained with softmax in your paper was 74.81%, while I retrained the baseline and got 77.45%. Does this mean the baseline result reported in your paper is too low?
![image](https://user-images.githubusercontent.com/18299981/62188302-87704f80-b39e-11e9-8fe8-61bb0e77b8b5.png)

**Comment by richardaecn** [comment]
I re-ran the code on CIFAR-10 (imbalance factor: 50) and achieved 75.98% for vanilla softmax and 79.18% for class-balanced softmax with beta=0.9999 with the last checkpoint. 

Compared with the reported numbers of 74.81% and 78.05%, the performances for both models seem improved. It could be due to the randomness of training the model or the different version of TF, CUDA and cuDNN. I ran the code with TF 1.14, CUDA 10.0 and cnDNN 7.6.1. 

However, the improvement of class-balanced loss is consistent - reported in the paper: 78.05% - 74.81% = 3.24%; new ran: 79.18% - 75.98% = 3.2%.

![Screenshot from 2019-07-31 00-10-13](https://user-images.githubusercontent.com/3731219/62192162-f2337380-b329-11e9-8780-c1a2338d2128.png)

**Comment by hbsz123** [comment]
> I re-ran the code on CIFAR-10 (imbalance factor: 50) and achieved 75.98% for vanilla softmax and 79.18% for class-balanced softmax with beta=0.9999 with the last checkpoint.
> 
> Compared with the reported numbers of 74.81% and 78.05%, the performances for both models seem improved. It could be due to the randomness of training the model or the different version of TF, CUDA and cuDNN. I ran the code with TF 1.14, CUDA 10.0 and cnDNN 7.6.1.
> 
> However, the improvement of class-balanced loss is consistent - reported in the paper: 78.05% - 74.81% = 3.24%; new ran: 79.18% - 75.98% = 3.2%.
> 
> ![Screenshot from 2019-07-31 00-10-13](https://user-images.githubusercontent.com/3731219/62192162-f2337380-b329-11e9-8780-c1a2338d2128.png)

Could you please also run the class balanced focal loss?
I also ran the class balanced focal loss and get 79.48% accuracy. Compared with the baseline softmax I ran(77.47%), class balanced focal loss has just 2% gain in performance.
![image](https://user-images.githubusercontent.com/18299981/62192811-e8048a00-b3a8-11e9-9d14-c4d5e6b2e9fd.png)

**Comment by richardaecn** [comment]
> Could you please also run the class balanced focal loss?
> I also ran the class balanced focal loss and get 79.48% accuracy. Compared with the baseline softmax I ran(77.47%), class balanced focal loss has just 2% gain in performance.
> ![image](https://user-images.githubusercontent.com/18299981/62192811-e8048a00-b3a8-11e9-9d14-c4d5e6b2e9fd.png)

Sure. I will re-ran and update to you soon. For the softmax baseline you got, can you also re-ran it multiple times (using first 10 lines in cifar_im_trainval.sh) and see if it's consistently achieving that high accuracy? On my side, the softmax baseline is usually 75-76%.

**Comment by hbsz123** [comment]
ok.
I will rerun for several times.
Thanks for your patientce. 


------------------ Original ------------------

**Comment by richardaecn** [comment]
> Could you please also run the class balanced focal loss?
> I also ran the class balanced focal loss and get 79.48% accuracy. Compared with the baseline softmax I ran(77.47%), class balanced focal loss has just 2% gain in performance.
> ![image](https://user-images.githubusercontent.com/18299981/62192811-e8048a00-b3a8-11e9-9d14-c4d5e6b2e9fd.png)

I got 79.42% with class-balanced focal loss (gamma = 1.0 and beta = 0.9999). I also updated README with tensorboard results.

**Comment by hbsz123** [comment]
OK, thanks.
The baseline  I ran with vanilla softmax is almost always above 76.5%.
![image](https://user-images.githubusercontent.com/18299981/62273287-c0c4c000-b46f-11e9-9d0b-3b4a1a11867c.png)

**Comment by hbsz123** [comment]
By the way, did you train iNaturalist2018 from scratch in your paper? 
Why not train with ImageNet pretrained model?

**Comment by richardaecn** [comment]
Yes, we train iNat from scratch. We did not fine-tune because iNat is large enough for training from scratch and ImageNet data is biased so we want to exclude the influence of ImageNet data. Another reason is that we want to use the same standard training strategy with ImageNet training.

**Comment by hbsz123** [comment]
Thanks for your reply.
The theory of effictive number in your paper is very impressive.
Do you have wechat？I have some other questions to ask about and I will close this issue.




------------------ Original ------------------
